tests/AssertAnnotationTest.php

```diff
@@ -2055,6 +2055,27 @@ function makeLowerNonEmpty(string $input): string
                         return $input;
                     }',
             ],
+            'assertListKeepsArrayShape' => [
+                'code' => '<?php
+                    /**
+                     * @param mixed $arr
+                     * @psalm-assert-if-true list $arr
+                     */
+                    function is_list($arr): bool
+                    {
+                        return is_array($arr) && array_is_list($arr);
+                    }
+
+                    /** @var array{0: int, 1: bool, 2: string} */
+                    $list = [1, true, "string"];
+                    assert(is_list($list));
+
+                    function takesString(string $_str): void {}
+
+                    takesString($list[2]);
+                ',
+                'assertions' => ['$list===' => "array{0: int, 1: bool, 2: string}<int<3, max>, mixed>"],
```

Comment:
Could you just check what is the resulting type if you have a `array{0: int, 1: bool, 2: string}<int<3, max>, mixed>` for $var and you do a `$var[] = 'a'`?

Reply:
It becomes `non-empty-list<bool|int|mixed|string>`, which seems a bit unfortunate since that's strictly less specific than the previous type... Would you prefer if I tried to fix that here or leave it for a future PR?